components/raftstore-v2/src/operation/command/admin/merge/mod.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
+
+pub mod prepare;
+
+use engine_traits::{KvEngine, RaftEngine};
+use kvproto::{
+    raft_cmdpb::{AdminCmdType, RaftCmdRequest},
+    raft_serverpb::{MergeState, PeerState, RegionLocalState},
+};
+use raft::{ProgressState, INVALID_INDEX};
+use raftstore::Result;
+use slog::{info, warn, Logger};
+use tikv_util::box_err;
+
+use crate::raft::Peer;
+
+pub enum PrepareStatus {
+    /// When a fence <idx, cmd> is present, we (1) delay the PrepareMerge
+    /// command `cmd` until all writes before `idx` are applied (2) reject all
+    /// in-coming write proposals.
+    /// Before proposing `PrepareMerge`, we first serialize and propose the lock
+    /// table. Locks marked as deleted (but not removed yet) will be
+    /// serialized as normal locks.
+    /// Thanks to the fence, we can ensure at the time of lock transfer, locks
+    /// are either removed (when applying logs) or won't be removed before
+    /// merge (the proposals to remove them are rejected).
+    ///
+    /// The request can be `None` because we needs to take it out to redo the
+    /// propose. In the meantime the fence is needed to bypass the check.
+    WaitForFence(u64, Option<RaftCmdRequest>),
+    /// In this state, all write proposals except for `RollbackMerge` will be
+    /// rejected.
+    Applied(MergeState),
+}
+
+#[derive(Default)]
+pub struct MergeContext {
+    pub prepare_status: Option<PrepareStatus>,
+}
+
+impl MergeContext {
+    #[inline]
+    pub fn from_region_state(logger: &Logger, state: &RegionLocalState) -> Option<Self> {
+        if state.get_state() == PeerState::Merging {
+            info!(logger, "region is merging"; "region_state" => ?state);
+            let mut ctx = Self::default();
+            ctx.prepare_status = Some(PrepareStatus::Applied(state.get_merge_state().clone()));
+            Some(ctx)
+        } else {
+            None
+        }
+    }
+
+    #[inline]
+    pub fn maybe_redo_pending_prepare(&mut self, applied: u64) -> Option<RaftCmdRequest> {
+        if let Some(PrepareStatus::WaitForFence(f, cmd)) = self.prepare_status.as_mut()
+            && applied >= *f
+        {
+            // The status will be updated during processing the proposal.
+            return cmd.take();
+        }
+        None
+    }
+
+    #[inline]
+    pub fn should_block_write(&self, admin_type: Option<AdminCmdType>) -> bool {
+        matches!(self.prepare_status, Some(PrepareStatus::Applied(_)))
```

Comment:
Why not use this instead? Proposal control can't check `WaitForFence` anyway.

Reply:
Because`ProposalControl` is designed to control proposals. It's better to use the same control API instead of keep adding new checks. Blocks to index can also be a supplementary for `ProposalControl`.

Reply:
I agree they are both controlling propose, but that's where similarity ends. The merge control doesn't need `delay_chs`, so we can't reuse `check_conflict`, it also needs to return custom error (`ProposalInMergingMode`), the "block till index" abstraction isn't clean either because it still only allows `PrepareMerge` after reaching the index.

Reply:
> The merge control doesn't need delay_chs, so we can't reuse check_conflict, it also needs to return custom error (ProposalInMergingMode),

Actually `delay_chs` can be also applied to merge as an optimization to reduce unnecessary backoff and retry. The custom error is not a hard requirement and it can be returned after the source peer actually enters merge.